0038-http-auth-profiles/http-auth-profiles.md

```diff
@@ -0,0 +1,183 @@
+---
+title: Interledger HTTP Authentication Profiles
+draft: 1
+---
+
+# Interledger HTTP Authentication Profiles
+A minimal set of extensible, token-based profiles for securing Interledger relationships using HTTP.
+
+## Motivation
+Many Interledger protocols and software work well using HTTP. For example, [ILP-over-HTTP](https://github.com/interledger/rfcs/blob/master/0035-ilp-over-http/0035-ilp-over-http.md) utilizes HTTP at the [Link Layer](https://interledger.org/rfcs/0001-interledger-architecture/). Additionally, Interledger software often uses HTTP for administrative functionality and other purposes.
+
+This document defines common profiles that can be used to ensure implementations of Interledger software work securely in an interoperable fashion, while simultaneously allowing Interledger software operators to more clearly understand the trade-offs involved with various security choices.
+
+The profiles defined in this RFC should:
+ 
+* ...clearly define any trade-offs involved in choosing a particular profile.
+* ...be easily implementable across programming languages in order to aid broad adoption.
+* ...utilise well-known and broadly deployed protocols in order to increase confidence in the profiles.
+* ...be suitable for high-performance use-cases, such as ILPv4 packet processing.
+* ...resist replays (emphasis: resist, not necessarily be replay-proof, in order to support high-performance use-cases).
+* ...not negate or prevent the ability of operators to layer-on more security "on top of" a profile defined in this RFC.
+
+## Specification
+
+### Overview
+This document defines three authentication profiles for securing HTTP network connections in Interledger.
+
+### Authentication
+When authenticating requests between Interledger nodes, it is important to choose an algorithm that maintains an appropriate balance between usability and security, while at the same time maintaining high-performance _and_ interoperability.
+
+In order to find this balance, this document defines three Authentication profiles, each with various trade-offs that should be considered before use:
+
+* `SIMPLE`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret as a [Bearer token](https://tools.ietf.org/html/rfc6750) in all HTTP requests. Peers SHOULD consider this token to be opaque and SHOULD NOT derive any special meaning from the token. 
```

Comment:
> So, perhaps we can describe SIMPLE as something else that is more along the lines of an "opaque token plus introspection" and then provide guidance on the issuance of the token such as setting an expiry?

Hmm....I like this angle, although I'm trying to figure out how to map it to RFC verbiage. Things like expiry are really a peer-wise decision, and I don't think I can make a reasonable recommendation here, especially since _most_ `SIMPLE` tokens will _never_ expire.

Also, recommending a specific deployment model seems a bit improper/premature for this RFC. If one set of peers want to use something like Hydra/Oathkeeper, and a different set of peers want to have the connector generate and validate auth tokens directly with no proxy, it seems like both should be valid choices assuming people know the tradeoffs for their use-case.

In my mind the only remaining question is: when using a SIMPLE auth token, how does the resource-server determine identity (whatever that means for the resource server). I _think_ the answer is that it doesn't matter for this RFC. Take the Rust connector as an example: it will concatenate a `UserId` + `Random bytes` to create a bearer token. Only it knows how to understand this token, but because it generated the token, this should be fine.

Reply:
> Things like expiry are really a peer-wise decision

I think the issue here is a "shared" token. Just because a token is opaque doesn't mean both sides should use the same token.

I think we should rather recommend that Alice gets a token from Bob and Bob decides how long it's valid and when to revoke it. Bob then also gets a token from Alice and Alice decides how long its valid and when to revoke it.

In many cases there is a client/server relationship (WebSockets) so it's okay  for just the server to issue a token to the client.

Are there implementations where ILP-over-HTTP is using the same token from both clients?

Reply:
> I think the issue here is a "shared" token. Just because a token is opaque doesn't mean both sides should use the same token.

Perhaps the proposed language can be improved here, but each of the profiles defined in this PR are "one-way", meaning that for any peer-relationship, either peer could choose a different profile. For example, were the Java connector to peer with the Rust connector, the Rust connector might be using `SIMPLE` for its tokens, whereas the Java Connector might only produce `JWT_HS_256` tokens. This type of flow should work fine because each "authenticating peer" is in charge of only accepting tokens it decides to accept.

From this perspective, the token is "shared" (i.e., if I'm making a request to you using the `SIMPLE` token you provided me, then we both know the token). But "shared" doesn't mean we have to use the same token in both directions.

> I think we should rather recommend that Alice gets a token from Bob and Bob decides how long it's valid and when to revoke it. Bob then also gets a token from Alice and Alice decides how long its valid and when to revoke it.

Yes, I agree - that was my intention with this (although if two peers want to use the same token in both directions, that's up to them). Let me see if I can edit `SIMPLE` to be a bit clearer here.

Reply:
I think all comments in this section have been addresses with the new `SIMPLE` profile wording. Resolving this here - make a comment inline if you have more thoughts.